### PR TITLE
DOC: Better docs for `jnp.convolve` & `jnp.correlate`

### PR DESCRIPTION
**Pull request by rajasekharporeddy**
part of #21461.

### CODE REVIEW

File: jax/_src/numpy/lax_numpy.py [hunk 1]
```diff
@@ -397,23 +397,131 @@ def _conv(x: Array, y: Array, mode: str, op: str, precision: PrecisionLike,
...
+    Array containing the convolved result.
+
+  See Also:
+    - :func:`jax.scipy.signal.convolve`: ND convolution
+    - :func:`jax.numpy.correlate`: 1D correlation
+
+  Examples:
+    A few 1D convolution examples:
+
+    >>> x = jnp.array([1, 2, 3, 2, 1])
```
Comment (jakevdp) [inline jax/_src/numpy/lax_numpy.py:436]: Using symmetric arrays for examples might obscure the fact that `convolve` reverses the first input

File: jax/_src/numpy/lax_numpy.py [hunk 2]
```diff
@@ -397,23 +397,131 @@ def _conv(x: Array, y: Array, mode: str, op: str, precision: PrecisionLike,
...
+  """Correlation of two one dimensional arrays.
+
+  JAX implementation of :func:`numpy.correlate`.
+
+  Args:
+    a: left-hand input to the correlation. Must have ``a.ndim == 1``.
+    v: right-hand input to the correlation. Must have ``v.ndim == 1``.
+    mode: controls the size of the output. Available operations are:
+
+      * ``"full"``: (default) output the full correlation of the inputs.
```
Comment (jakevdp) [inline jax/_src/numpy/lax_numpy.py:475]: `correlate` defaults to `"valid"`

File: jax/_src/numpy/lax_numpy.py [hunk 3]
```diff
@@ -397,23 +397,131 @@ def _conv(x: Array, y: Array, mode: str, op: str, precision: PrecisionLike,
...
+    - :func:`jax.scipy.signal.correlate`: ND correlation
+    - :func:`jax.numpy.convolve`: 1D convolution
+
+  Examples:
+    >>> x = jnp.array([1, 2, 3, 2, 1])
+    >>> y = jnp.array([4, 5, 6])
+
+    Full correlation uses implicit zero-padding at the edges.
+
+    >>> jnp.correlate(x, y, mode='full')
```
Comment (jakevdp) [inline jax/_src/numpy/lax_numpy.py:501]: Show the default example first?

File: jax/_src/numpy/lax_numpy.py (hunk 1 above)
Comment (rajasekharporeddy) [reply jax/_src/numpy/lax_numpy.py:436]: Thanks! Changed `y` to non-symmetric array.

File: jax/_src/numpy/lax_numpy.py (hunk 2 above)
Comment (rajasekharporeddy) [reply jax/_src/numpy/lax_numpy.py]: Sorry, I didn't observe this. Updated!

File: jax/_src/numpy/lax_numpy.py (hunk 3 above)
Comment (rajasekharporeddy) [reply jax/_src/numpy/lax_numpy.py:508]: Done

File: jax/_src/numpy/lax_numpy.py [hunk 4]
```diff
@@ -397,23 +397,132 @@ def _conv(x: Array, y: Array, mode: str, op: str, precision: PrecisionLike,
...
+    - :func:`jax.numpy.convolve`: 1D convolution
+
+  Examples:
+    >>> x = jnp.array([1, 2, 3, 2, 1])
+    >>> y = jnp.array([4, 5, 6])
+
+    Since default ``mode = 'valid'``, ``jax.numpy.correlate`` returns only the
+    portion of correlation where the two arrays fully overlap:
+
+    >>> jax.scipy.signal.convolve(x, y)
```
Comment (jakevdp) [inline jax/_src/numpy/lax_numpy.py:502]: correlate

Comment (rajasekharporeddy) [reply jax/_src/numpy/lax_numpy.py]: Changed.

File: jax/_src/numpy/lax_numpy.py [hunk 5]
```diff
@@ -397,23 +397,132 @@ def _conv(x: Array, y: Array, mode: str, op: str, precision: PrecisionLike,
...
 
-@util.implements(np.convolve, lax_description=_PRECISION_DOC,
-                 extra_params=_CONV_PREFERRED_ELEMENT_TYPE_DESCRIPTION)
 @partial(jit, static_argnames=('mode', 'precision', 'preferred_element_type'))
 def convolve(a: ArrayLike, v: ArrayLike, mode: str = 'full', *,
              precision: PrecisionLike = None,
              preferred_element_type: DTypeLike | None = None) -> Array:
+  """Convolution of two one dimensional arrays.
+
+  JAX implementation of :func:`numpy.convolve`.
```
Comment (jakevdp) [inline jax/_src/numpy/lax_numpy.py:406]: I'm putting myself in the user's shoes, and I think I'd still be confused about what this does (particularly in contrast to `jnp.correlate`).

Maybe we should add a mathematical definition, something like
```
$c_k = \sum_j a_{k - j} v_j$
```
which renders as

$$
c_k = \sum_j a_{k - j} v_j
$$

I think this is the correct expression for the full 1D convolution.

The expression for correlation is similar, I think, except it's `k + j` and `v` is conjugated.

That would make it more clear at a glance what the relationship is between the two functions.

What do you think?

Comment (rajasekharporeddy) [reply jax/_src/numpy/lax_numpy.py:406]: Yes, providing mathematical definition would be great in user point of view. Included the definition. Thanks @jakevdp.

Comment (rajasekharporeddy) [reply jax/_src/numpy/lax_numpy.py:406]: Also included example for complex arrays.

Comment (rajasekharporeddy) [reply jax/_src/numpy/lax_numpy.py:406]: I have a doubt on the type of inputs. Does they need to be `Array` or `ArrayLike`?

Comment (jakevdp) [reply jax/_src/numpy/lax_numpy.py:406]: Inputs should be `ArrayLike`, which does not include lists.

Comment (rajasekharporeddy) [reply jax/_src/numpy/lax_numpy.py:406]: Understood. Thanks for clarifying.

**Comment by rajasekharporeddy** [comment]
If no other changes are required, I would squash all commits.

**Comment by rajasekharporeddy** [comment]
Hi @jakevdp, Could you please start the tests once again. Earlier some tests failed since I haven't used raw triple double quotes (r""" """).Now I fixed it.

**Comment by jakevdp** [comment]
Looks good - once this is squashed we can merge.

**Comment by rajasekharporeddy** [comment]
Squashed into single commit.

**Review by jakevdp — APPROVED** [review]
Looks great, thanks!

**Pull request merged**